lib/time/time.go

```diff
@@ -0,0 +1,462 @@
+	}
+	loc, err := time.LoadLocation(string(x))
+	if err != nil {
+		return nil, err
+	}
+
+	return starlark.String(loc.String()), nil
+}
+
+func sleep(thread *starlark.Thread, _ *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var x starlark.Int
+	if err := starlark.UnpackArgs("sleep", args, kwargs, "x", &x); err != nil {
+		return starlark.None, err
+	}
+	dur, ok := x.Int64()
+	if !ok {
+		return starlark.None, fmt.Errorf("invalid sleep value")
+	}
+
+	time.Sleep(time.Duration(dur))
+	return starlark.None, nil
+}
+
+func parseTime(thread *starlark.Thread, _ *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var (
+		x, location starlark.String
+		format      = starlark.String(time.RFC3339)
+	)
+	if err := starlark.UnpackArgs("time", args, kwargs, "x", &x, "format?", &format, "location", &location); err != nil {
+		return nil, err
+	}
+
+	if location == "" {
+		t, err := time.Parse(format.String(), x.String())
+		if err != nil {
+			return nil, err
+		}
+		return Time(t), nil
+	}
+
+	loc, err := time.LoadLocation(location.String())
+	if err != nil {
+		return nil, err
+	}
+	t, err := time.ParseInLocation(format.String(), x.String(), loc)
+	if err != nil {
+		return nil, err
+	}
+	return Time(t), nil
+}
+
+func fromUnixTime(thread *starlark.Thread, _ *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var (
+		x starlark.Int
+	)
+	if err := starlark.UnpackArgs("time", args, kwargs, "x", &x); err != nil {
+		return nil, err
+	}
+
+	i, err := strconv.ParseInt(x.String(), 10, 64)
+	if err != nil {
+		return nil, err
+	}
+
+	t := time.Unix(i, 0)
+
+	return Time(t), nil
+}
+
+func now(thread *starlark.Thread, _ *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	return Time(NowFunc()), nil
+}
+
+// Duration is a starlark representation of a duration.
+type Duration time.Duration
+
+// String implements the Stringer interface.
+func (d Duration) String() string { return time.Duration(d).String() }
+
+// Type returns a short string describing the value's type.
+func (d Duration) Type() string { return "time.duration" }
+
+// Freeze renders Duration immutable. required by starlark.Value interface
+// because duration is already immutable this is a no-op.
+func (d Duration) Freeze() {}
+
+// Hash returns a function of x such that Equals(x, y) => Hash(x) == Hash(y)
+// required by starlark.Value interface.
+func (d Duration) Hash() (uint32, error) { return starlark.MakeInt64(int64(d)).Hash() }
+
+// Truth returns the truth value of an object required by starlark.Value
+// interface.
+func (d Duration) Truth() starlark.Bool { return d != 0 }
+
+// Attr gets a value for a string attribute, implementing dot expression support
+// in starklark. required by starlark.HasAttrs interface.
+func (d Duration) Attr(name string) (starlark.Value, error) {
+	return builtinAttr(d, name, durationMethods)
```

Comment:
Methods are appropriate when there is potential computation or failure, but for the component accessors of a duration or time, fields are more appropriate.